### ISSUE TITLE
[NEED ASSIST] encryptToken function missing parameter

### ISSUE THREAD

**Issue by mkcpb** [Adyen/adyen-ios#1232]
**Describe the challenge**
Hi, I updated Adyen SDK to version 5.1.0. I tested if everything works as expected and when I enter card details I get an error that holder name is not provided. I found that missing .add(holder: card.holder) in this function:

public static func encryptToken(from card: Card, with publicKey: String) throws -> String {
        guard !card.isEmpty else { throw CardEncryptor.Error.invalidCard }
        let payload = CardPayload()
              .add(number: card.number)
              .add(securityCode: card.securityCode)
              .add(expiryMonth: card.expiryMonth)
              .add(expiryYear: card.expiryYear)
        return try encrypt(payload, with: publicKey)
}

When I add that line it works as expected, so I think you need to add that. Maybe I miss something?
The point is that we have to migrate because the old checkout will soon no longer work and we should implement a drop in or component. For a while, we will have to use both, the old version and the new solution that will be implemented until we completely switch to the new solution.

**Comment by erenbesel** [comment]
Hello @mkcpb , thanks for reaching out!

Holder name is not a field that needs to be encrypted. Are you trying to encrypt that property?
How do you create the card component?

I tested on our demo app with enabling the holder name option in the `CardComponent.Configuration` and it worked fine.
```
/// Indicates if the field for entering the holder name should be displayed in the form. Defaults to false.
public var showsHolderNameField: Bool
```

**Comment by mkcpb** [comment]
Hi @erenbesel, yes I try to encrypt holdername property. How it looks like before: 
```
final class AdyenPaymentEncryptUtils: NSObject {
    
    class func enryptCardInfo(number: String,
                              holderName: String,
                              cvc: String,
                              expiryMonth: String,
                              expiryYear: String) throws -> String? {
        // Create a card object.
        let card = CardEncryptor.Card(number: number, securityCode: cvc, expiryMonth: expiryMonth, expiryYear: expiryYear)
        
        if let publicKey = Bundle.main.object(forInfoDictionaryKey: "AdyenAppKey") as? String {
            
            // Encrypt card data.
            do {
                return try CardEncryptor.encryptedToken(for: card, holderName: holderName, publicKey: publicKey)
            } catch {
                throw AdyenEncryptionError.cardEncodingError
            }
            
        } else {
            
            fatalError("Check ADYEN_APP_KEY Configuration")
        }
    }
}
```
How I think should looks like now after SDK update: 
```
final class AdyenPaymentEncryptUtils: NSObject {
    
    class func enryptCardInfo(number: String,
                              holderName: String,
                              cvc: String,
                              expiryMonth: String,
                              expiryYear: String) throws -> String? {
        // Create a card object.
        
        let card = Card(number: number, securityCode: cvc, expiryMonth: expiryMonth, expiryYear: expiryYear, holder: holderName)

        if let publicKey = Bundle.main.object(forInfoDictionaryKey: "AdyenAppKey") as? String {

            // Encrypt card data.
            do {
                return try CardEncryptor.encryptToken(from: card, with: publicKey)
            } catch {
                throw AdyenEncryptionError.cardEncodingError
            }

        } else {

            fatalError("Check ADYEN_APP_KEY Configuration")
        }
    }
}
```

**Comment by mkcpb** [comment]
Hi @erenbesel, so to get encrypted card as a token I think the holder name should be encrypted also in order to classic still checkout works. We used `CardEncryptor.encryptedToken(for: card, holderName: holderName, publicKey: publicKey)` so I think we still need that. What do you think?

**Comment by mkcpb** [comment]
Hi @erenbesel , sorry but I really need help here. I'm blocked with this. We used Adyen 3.7.0 version. This is code from 3.7.0 version:
```
/// Encrypts a card.
    ///
    /// - Parameters:
    ///   - card: Card containing the data to be encrypted.
    ///   - holderName: The cardholder's name.
    ///   - publicKey: The public key to use for encryption (format "Exponent|Modulus").
    /// - Returns: A string representing the encrypted card.
    /// - Throws:  `CardEncryptor.Error.encryptionFailed` if the encryption failed,
    ///  maybe because the card public key is an invalid one, or for any other reason.
    /// - Throws:  `CardEncryptor.Error.invalidEncryptionArguments` when trying to encrypt a card with  card number, securityCode,
    /// expiryMonth, expiryYear, and holderName, all of them are nil.
    /// - Throws:  `CardEncryptor.Error.unknown` if encryption failed for an unknown reason.
    @available(*, deprecated, message: "Use `card.encryptedToToken(publicKey:holderName:)`.")
    public static func encryptedToken(for card: Card, holderName: String?, publicKey: String) throws -> String {
        guard let cardToken = try card.encryptedToToken(publicKey: publicKey, holderName: holderName) else {
            // This should never happen,
            // because card.encryptedToToken(publicKey:holderName:) will throw an error if the generated token is some how nil,
            // before reaching here.
            throw Error.unknown
        }
        
        return cardToken
    } 
```
```
func encryptedToToken(publicKey: String, holderName: String?) throws -> String? { // swiftlint:disable:this explicit_acl
        // Make sure not all the card ivars's are nil, otherwise throw Error.invalidEncryptionArguments
        guard [number, securityCode, expiryMonth, expiryYear, holderName].contains(where: { $0 != nil }) else {
            throw CardEncryptor.Error.invalidEncryptionArguments
        }
        guard let encryptionResult = ObjC_CardEncryptor.encryptedToToken(withNumber: number,
                                                                         securityCode: securityCode,
                                                                         expiryMonth: expiryMonth,
                                                                         expiryYear: expiryYear,
                                                                         holderName: holderName,
                                                                         publicKey: publicKey) else {
            throw CardEncryptor.Error.encryptionFailed
            
        }
        return encryptionResult
    }
```

**Comment by descorp** [comment]
Hey @mkcpb 

Our v4 and v5 SDK were never tested with [Classic PAL API](https://docs.adyen.com/api-explorer/Payment/68/overview).
Is it possible for you to switch to a [Checkout API](https://docs.adyen.com/api-explorer/Checkout/68/overview)?

**Comment by mkcpb** [comment]
Hi @descorp,

We are switching now but as I said before the thing is that we need to support classic checkout some time also. Our plan is to migrate classic checkout to drop-in and some period of time to have both options(for a small number of users we will use the new solution and monitor the behavior) until we are sure everything is working. So we used v3.7.0. From version v3.8.0 classic checkout is not working for us. To implement drop-in Adyen version should be at least 3.8.0.

**Comment by descorp** [comment]
> Our plan is to migrate classic checkout to drop-in .. we need to support classic checkout some time also. 

Are you looking to use one version of SDK to work for both Classic API and Checkout API ?

You can copy our [v3.7.0 Encryption](https://github.com/Adyen/adyen-ios/tree/3.7.0/AdyenCard/Utilities/Encryption) code locally for Classic API and use `v5.2.0` SDK version for DropIn with Checkout API.

> To implement drop-in Adyen version should be at least 3.8.0.

What is the issue with 3.7.0? One can use DropInComponent [on 3.7.0 as well](https://github.com/Adyen/adyen-ios/tree/3.7.0/AdyenDropIn).

**Comment by mkcpb** [comment]
Hey @descorp ,

Yes, exactly that. We started migrating because Classic Checkout SDK is end-of-life as described here: [](https://docs.adyen.com/online-payments/classic-integrations/checkout-sdks/deprecation#migration-steps). The new implementation would be used by a smaller number of users to avoid possible problems. We implement a mechanism so that we can immediately revert to the old implementation if a problem occurs. 

If I can use 3.7.0 then I don't have to update the version. But as you can see here version should be at least 3.8.0. So is it true or not?

<img width="701" alt="Screenshot 2023-06-07 at 16 12 45" src="https://github.com/Adyen/adyen-ios/assets/92429728/c9573bd9-7b65-408e-b4b0-7e13e4d4c11a">

**Comment by descorp** [comment]
SPM was introduced on 3.8.0 - it became possible after rewriting the Encryption code to Swift.
To access versions prior to 3.8.0 - use Cocoapods or Carthage.

**Comment by mkcpb** [comment]
Ah ok, then I misunderstood. I already use Cocoapods. I apologize and thank you for your time.

**Comment by mkcpb** [comment]
Hi @descorp, @erenbesel , I come back on this. I'm little bit confused about this. So, For example if I try to setting up the Drop-in on v3.7.0 I don't have an instance AdyenContext which is located in Adyen/Actions so I need to update version to v3.8.0+. For example if I use 4.1.0 version and I run the project I get an error message in AdyenNetworking(version 1.0.0). When I update Adyen version to 5.2.0 and AdyenNetworking to 2.0.0 I can run the project without errors. So, on our backend is implemented three api calls: payment-methods, authorize and additional-details but as I can see in documentation we can use these calls only to implement Drop-in versions between 4.9.0 - 4.10.3. For the version 5.2.0 we need to use sessions api call. So, is it possible to implement Drop-in version 5.2.0 using three api calls that we implemented? Did you tested Drop-in 4.9.0 - 4.10.3 implementation? As I can see I can't proceed with any of these versions, am I right?

**Comment by descorp** [comment]
Hey @mkcpb 

> For the version 5.2.0 we need to use sessions api

One can still use 3 endpoints (aka ["Advanced flow"](https://docs.adyen.com/online-payments/build-your-integration/additional-use-cases/advanced-flow-integration?platform=iOS&integration=Drop-in&version=5.1.0)) on v5
